src/libaktualizr/utilities/dequeue_buffer.cc

```diff
@@ -0,0 +1,50 @@
+#include "utilities/dequeue_buffer.h"
+
+#include <cassert>
+#include <cstring>
+#include <stdexcept>
+char* DequeueBuffer::Head() {
+  assert(sentinel_ == kSentinel);
+  return buffer_.data();
+}
+
+size_t DequeueBuffer::Size() {
+  assert(sentinel_ == kSentinel);
+  return written_bytes_;
+}
+
+void DequeueBuffer::Consume(size_t bytes) {
+  assert(sentinel_ == kSentinel);
+  // It would be possible to have a smarter algorithm here that
+  // only shuffles bytes down when Tail() is called and we are getting
+  // close to the end of the buffer, or when the buffer is nearly empty
+  // the memmove operation is cheap. This isn't performance critical code.
+  if (written_bytes_ < bytes) {
+    throw std::logic_error("Attempt to DequeueBuffer::Consume() more bytes than are valid");
+  }
+  size_t remaining = written_bytes_ - bytes;
+  std::copy_backward(buffer_.begin() + bytes, buffer_.begin() + written_bytes_, buffer_.begin() + remaining);
```

Comment:
Why `copy_backward`? Doesn't that risk undefined/dangerous behavior if `bytes < remaining`?

Reply:
[std::move](http://en.cppreference.com/w/cpp/algorithm/move) is overloaded apparently

Reply:
[std::copy](http://en.cppreference.com/w/cpp/algorithm/copy) should be fine, too.

Reply:
It has to be one of the _backward interfaces, because (unlike `memmove`) `std::copy` and `std::move` are undefined in the cases where the input and output ranges overlap.

I much prefer the `memmove` API, but I wanted to stick to the STL where possible.

Reply:
sorry, you are right-this should be std::copy after all